File: fixtures/rust-sample/src/spawn.rs
// cgx-fixture: spawn edges and detached error domain
// Covers: tokio::spawn -> spawns edge (GM-9),
//         detached error domain (seen_exceptional reset at spawns edge),
//         spawns edge condition (always/conditional/loop)

use std::sync::{Arc, Mutex};

async fn background_work(id: u32) {
    process_task(id).await;
}

async fn process_task(id: u32) {
    log_task(id);
}

fn log_task(id: u32) {
    println!("task {}", id);
}

fn handle_join_error(e: tokio::task::JoinError) {
    eprintln!("task failed: {}", e);
}

/// Unconditional spawn — spawns edge with edge_condition=always.
pub async fn spawn_one() {
    let _handle = tokio::spawn(background_work(1));  // spawns edge, always
}

/// Conditional spawn — spawns edge with edge_condition=conditional.
pub async fn spawn_if(condition: bool) {
    if condition {
        tokio::spawn(background_work(2));  // spawns edge, conditional
    }
}

/// Loop spawn — spawns edge with edge_condition=loop.
pub async fn spawn_many(count: u32) {
    for i in 0..count {
        tokio::spawn(background_work(i));  // spawns edge, loop
    }
}

/// Spawn in error handling context — spawns edge with edge_condition=exception.
/// Demonstrates that the detached task's domain is independent.
pub async fn spawn_on_error(result: Result<u32, String>) {
    match result {
        Ok(id) => process_task(id).await,
        Err(_) => {
            tokio::spawn(background_work(0));  // spawns edge, exception (Err arm)
        }
    }
}

/// Using a shared Arc<Mutex<>> — demonstrates lock-set tracking context.
pub async fn spawn_with_shared_state(counter: Arc<Mutex<u32>>) {
    tokio::spawn(async move {
        let mut guard = counter.lock().unwrap();
        *guard += 1;
    });  // spawns edge; lock acquired inside spawned task, not in spawner
}

/// Join handle awaited — this is calls:async, NOT spawns.
pub async fn spawn_and_join() -> Result<(), String> {
    let handle = tokio::spawn(background_work(99));
    handle.await.map_err(|e| {
        handle_join_error(e);
        "join failed".to_string()
    })
}

File: fixtures/ts-sample/src/spawn.ts
// cgx-fixture: spawn and Promise patterns (TypeScript)
// Covers: Promise constructor -> spawns edge (detached execution),
//         setTimeout/setInterval -> spawns edge,
//         Promise.race/all -> calls:async,
//         detached error domain (GM-9)

async function backgroundWork(id: number): Promise<void> {
    await processTask(id);
}

async function processTask(id: number): Promise<void> {
    logTask(id);
}

function logTask(id: number): void {
    console.log(`task ${id}`);
}

function handleError(e: Error): void {
    console.error("task failed:", e.message);
}

/// Unconditional spawn via Promise + no await — spawns edge, always.
export function spawnOne(): void {
    void backgroundWork(1);   // spawns edge, always (fire-and-forget)
}

/// Conditional spawn — spawns edge, conditional.
export function spawnIf(condition: boolean): void {
    if (condition) {
        void backgroundWork(2);  // spawns edge, conditional
    }
}

/// Loop spawn — spawns edge, loop.
export function spawnMany(count: number): void {
    for (let i = 0; i < count; i++) {
        void backgroundWork(i);  // spawns edge, loop
    }
}

/// Spawn via setTimeout — spawns edge (detached timer callback).
export function spawnDelayed(ms: number): void {
    setTimeout(() => {
        void processTask(99);  // spawns edge, always (inside timer callback)
    }, ms);
}

/// setInterval — spawns edge, loop (repeated firing).
export function spawnInterval(ms: number): ReturnType<typeof setInterval> {
    return setInterval(() => {
        logTask(0);  // spawns edge via interval
    }, ms);
}

/// Spawn in error handling context — spawns edge, exception.
export function spawnOnError(result: Result<number, Error>): void {
    if (!result.ok) {
        void backgroundWork(0);  // spawns edge, conditional (error branch)
    }
}

type Result<T, E> = { ok: true; value: T } | { ok: false; error: E };

/// Awaited Promise — calls:async, NOT spawns.
export async function spawnAndJoin(): Promise<void> {
    await backgroundWork(99);   // calls:async (awaited — retains result)
}

/// Promise.all — awaited results from multiple concurrent tasks.
export async function allParallel(ids: number[]): Promise<void[]> {
    return Promise.all(ids.map(id => backgroundWork(id)));  // calls:async on the await
}
